File: Difficulty: Medium/Allocate Minimum Pages/allocate-minimum-pages.cpp
class Solution {
public:
    bool is_feasible(const std::vector<int>& arr, int k, int max_pages) {
        int students = 1, current_pages = 0;
        for (int pages : arr) {
            if (current_pages + pages > max_pages) {
                students++;
                current_pages = pages;
                if (students > k) return false;
            } else {
                current_pages += pages;
            }
        }
        return true;
    }

    int findPages(std::vector<int>& arr, int k) {
        if (arr.size() < k) return -1;

        int left = *std::max_element(arr.begin(), arr.end());
        int right = std::accumulate(arr.begin(), arr.end(), 0);
        int result = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (is_feasible(arr, k, mid)) {
                result = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return result;
    }
};

File: Difficulty: Medium/Allocate Minimum Pages/allocate-minimum-pages.java
class Solution {
    public int findPages(int[] arr, int k) {
        int sum=0,mx=0;
        if(k>arr.length) return -1;
        int n= arr.length;
        for(int i=0;i<arr.length;i++){
            sum+=arr[i];
            mx = Math.max(mx,arr[i]);
        }
        int low = mx, high = sum,res=0;
        while(low<=high){
            int mid  = (low+high)/2;
            if(isfeasible(arr,n,k,mid)){
                res =mid;
                high=mid-1;
            }else{
                low=mid+1;
            }
        }
        return res;
    }
    boolean isfeasible(int[] arr, int n, int k, int ans){
        n = arr.length;
        int req=1,sum=0;
        for(int i=0;i<n;i++){
            if(sum+arr[i]>ans){
                req++;
                sum=arr[i];
            }
            else{
                sum+=arr[i];
            }
        }
        return(req<=k);
    }
}